calling ds-test/test.sol's `failed()` results in 'NoneType' object has no attribute 'size'

**Describe the bug**

with halmos 0.1.14.dev15+g060a6d8.d20240813:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.15;

import "forge-std/Test.sol";
import {console2} from "forge-std/console2.sol";

contract Test91 is Test {
    function test_dstestFailed() external {
        console2.log("failed() =", failed());
    }
}
```

```shell
halmos --function test_dstestFailed --debug

Executing test_dstestFailed
Address 0x7109709ecfa91a80626ff3989d68f67f5b1dd12d not in: [0x7fa9385be102ac3eac297483dd6233d62b3e1496]
Empty address: 0x7109709ecfa91a80626ff3989d68f67f5b1dd12d
Address 0x7109709ecfa91a80626ff3989d68f67f5b1dd12d not in: [0x7fa9385be102ac3eac297483dd6233d62b3e1496]
Empty address: 0x7109709ecfa91a80626ff3989d68f67f5b1dd12d
Address 0x7109709ecfa91a80626ff3989d68f67f5b1dd12d not in: [0x7fa9385be102ac3eac297483dd6233d62b3e1496]
Empty address: 0x7109709ecfa91a80626ff3989d68f67f5b1dd12d
[ERROR] test_dstestFailed()
AttributeError: 'NoneType' object has no attribute 'size'
Traceback (most recent call last):
  File "/Users/karma/projects/halmos/src/halmos/__main__.py", line 1003, in run_sequential
    test_result = run(setup_ex, run_args.abi, fun_info, test_config)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/karma/projects/halmos/src/halmos/__main__.py", line 727, in run
    for idx, ex in enumerate(exs):
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 2651, in run
    self.call(ex, opcode, to_alias, stack, step_id)
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 2064, in call
    call_unknown()
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 1995, in call_unknown
    ret = hevm_cheat_code.handle(self, ex, arg, stack, step_id)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/karma/projects/halmos/src/halmos/cheatcodes.py", line 542, in handle
    return ByteVec(sevm.sload(ex, load_account_alias, load_slot))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 1679, in sload
    return self.storage_model.load(ex, addr, loc)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 1297, in load
    cls.init(ex, addr, loc)
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 1290, in init
    assert_address(addr)
  File "/Users/karma/projects/halmos/src/halmos/utils.py", line 478, in assert_address
    elif x.size() != 160:
         ^^^^^^
AttributeError: 'NoneType' object has no attribute 'size'
Symbolic test result: 0 passed; 1 failed; time: 0.03s

[time] total: 1.28s (build: 0.71s, load: 0.53s, tests: 0.04s)
```

`failed()` does the following:

```solidity
    function failed() public returns (bool) {
        if (_failed) {
            return _failed;
        } else {
            bool globalFailed = false;
            if (hasHEVMContext()) {  // <- true since we return extcodesize 1 for the hevm address
                (, bytes memory retdata) = HEVM_ADDRESS.call(
                    abi.encodePacked(
                        bytes4(keccak256("load(address,bytes32)")),
                        abi.encode(HEVM_ADDRESS, bytes32("failed"))
                    )
                );
                globalFailed = abi.decode(retdata, (bool));  // <- fails during the handling of vm.load(hevm, 'failed')
            }
            return globalFailed;
        }
    }
```

In halmos, vm.load is handled like this:

```python
        # vm.load(address,bytes32)
        elif funsig == hevm_cheat_code.load_sig:
            load_account = uint160(arg.get_word(4))  #  == 0x7109709ecfa91a80626ff3989d68f67f5b1dd12d
            load_slot = uint256(arg.get_word(36))
            load_account_alias = sevm.resolve_address_alias(
                ex, load_account, stack, step_id, branching=False
            )

            return ByteVec(sevm.sload(ex, load_account_alias, load_slot))
```

resolve_address_alias returns `None` for `load_account=0x7109709ecfa91a80626ff3989d68f67f5b1dd12d`, which leads to the downstream exception in `sevm.sload`

--- a/src/halmos/cheatcodes.py
+++ b/src/halmos/cheatcodes.py
@@ -354,6 +354,18 @@ class hevm_cheat_code:
         )
     )
 
+    # abi.encodePacked(
+    #     bytes4(keccak256("load(address,bytes32)")),
+    #     abi.encode(HEVM_ADDRESS, bytes32("failed"))
+    # )
+    failed_payload = ByteVec(
+        bytes.fromhex(
+            "667f9d70"
+            + "0000000000000000000000007109709ecfa91a80626ff3989d68f67f5b1dd12d"
+            + "6661696c65640000000000000000000000000000000000000000000000000000"
+        )
+    )
+
     # bytes4(keccak256("assume(bool)"))
     assume_sig: int = 0x4C63E562
 
@@ -533,6 +545,9 @@ class hevm_cheat_code:
 
         # vm.load(address,bytes32)
         elif funsig == hevm_cheat_code.load_sig:
+            if arg == hevm_cheat_code.failed_payload:
+                return ByteVec(con(0))
+
             load_account = uint160(arg.get_word(4))
             load_slot = uint256(arg.get_word(36))
             load_account_alias = sevm.resolve_address_alias(

--- a/src/halmos/cheatcodes.py
+++ b/src/halmos/cheatcodes.py
@@ -545,7 +545,10 @@ class hevm_cheat_code:
 
         # vm.load(address,bytes32)
         elif funsig == hevm_cheat_code.load_sig:
+            # vm.load(HEVM_ADDRESS, "failed") is handled separately
             if arg == hevm_cheat_code.failed_payload:
+                # since fail(), which triggers vm.store(HEVM_ADDRESS, "failed", 1), halts immediately, (see vm.store() above)
+                # the "failed" slot is never assigned, thus vm.load(HEVM_ADDRESS, "failed") always return zero at this point
                 return ByteVec(con(0))
 
             load_account = uint160(arg.get_word(4))

--- a/src/halmos/cheatcodes.py
+++ b/src/halmos/cheatcodes.py
@@ -208,6 +208,11 @@ def symbolic_storage(ex, arg, sevm, stack, step_id):
     account_alias = sevm.resolve_address_alias(
         ex, account, stack, step_id, branching=False
     )
+
+    if account_alias is None:
+        error_msg = f"enableSymbolicStorage() is not allowed for a nonexistent account: {hexify(account)}"
+        raise HalmosException(error_msg)
+
     ex.storage[account_alias].symbolic = True
 
 
@@ -351,18 +356,6 @@ class hevm_cheat_code:
             + "0000000000000000000000007109709ecfa91a80626ff3989d68f67f5b1dd12d"
             + "6661696c65640000000000000000000000000000000000000000000000000000"
             + "0000000000000000000000000000000000000000000000000000000000000001"
-        )
-    )
-
-    # abi.encodePacked(
-    #     bytes4(keccak256("load(address,bytes32)")),
-    #     abi.encode(HEVM_ADDRESS, bytes32("failed"))
-    # )
-    failed_payload = ByteVec(
-        bytes.fromhex(
-            "667f9d70"
-            + "0000000000000000000000007109709ecfa91a80626ff3989d68f67f5b1dd12d"
-            + "6661696c65640000000000000000000000000000000000000000000000000000"
         )
     )
 
@@ -540,22 +533,25 @@ class hevm_cheat_code:
                 ex, store_account, stack, step_id, branching=False
             )
 
+            if store_account_alias is None:
+                error_msg = f"vm.store() is not allowed for a nonexistent account: {hexify(store_account)}"
+                raise HalmosException(error_msg)
+
             sevm.sstore(ex, store_account_alias, store_slot, store_value)
             return ret
 
         # vm.load(address,bytes32)
         elif funsig == hevm_cheat_code.load_sig:
-            # vm.load(HEVM_ADDRESS, "failed") is handled separately
-            if arg == hevm_cheat_code.failed_payload:
-                # since fail(), which triggers vm.store(HEVM_ADDRESS, "failed", 1), halts immediately, (see vm.store() above)
-                # the "failed" slot is never assigned, thus vm.load(HEVM_ADDRESS, "failed") always return zero at this point
-                return ByteVec(con(0))
-
             load_account = uint160(arg.get_word(4))
             load_slot = uint256(arg.get_word(36))
             load_account_alias = sevm.resolve_address_alias(
                 ex, load_account, stack, step_id, branching=False
             )
+
+            if load_account_alias is None:
+                # since load_account doesn't exist, its storage is empty.
+                # note: the storage cannot be symbolic, as the symbolic storage cheatcode fails for nonexistent addresses.
+                return ByteVec(con(0))
 
             return ByteVec(sevm.sload(ex, load_account_alias, load_slot))
 

--- a/tests/expected/all.json
+++ b/tests/expected/all.json
@@ -1024,6 +1024,15 @@
                 "num_paths": null,
                 "time": null,
                 "num_bounded_loops": null
+            },
+            {
+                "name": "check_failed_cheatcode()",
+                "exitcode": 0,
+                "num_models": 0,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
             }
         ],
         "test/Foundry.t.sol:FoundryTest": [

--- a/tests/expected/all.json
+++ b/tests/expected/all.json
@@ -1246,6 +1246,15 @@
                 "num_paths": null,
                 "time": null,
                 "num_bounded_loops": null
+            },
+            {
+                "name": "check_enableSymbolicStorage()",
+                "exitcode": 3,
+                "num_models": 0,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
             }
         ],
         "test/Library.t.sol:LibraryTest": [
@@ -3151,6 +3160,15 @@
         ],
         "test/Store.t.sol:StoreTest": [
             {
+                "name": "check_load_nonexistent_account(uint256)",
+                "exitcode": 0,
+                "num_models": 0,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
+            },
+            {
                 "name": "check_store_Array(uint256,uint256)",
                 "exitcode": 0,
                 "num_models": 0,
@@ -3171,6 +3189,15 @@
             {
                 "name": "check_store_Scalar(uint256)",
                 "exitcode": 0,
+                "num_models": 0,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
+            },
+            {
+                "name": "check_store_nonexistent_account()",
+                "exitcode": 3,
                 "num_models": 0,
                 "models": null,
                 "num_paths": null,

--- a/tests/expected/all.json
+++ b/tests/expected/all.json
@@ -1248,8 +1248,26 @@
                 "num_bounded_loops": null
             },
             {
-                "name": "check_enableSymbolicStorage()",
+                "name": "check_enableSymbolicStorage_fail()",
+                "exitcode": 1,
+                "num_models": 1,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
+            },
+            {
+                "name": "check_enableSymbolicStorage_nonexistent()",
                 "exitcode": 3,
+                "num_models": 0,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
+            },
+            {
+                "name": "check_enableSymbolicStorage_pass(uint256)",
+                "exitcode": 0,
                 "num_models": 0,
                 "models": null,
                 "num_paths": null,

--- a/tests/regression/test/Foundry.t.sol
+++ b/tests/regression/test/Foundry.t.sol
@@ -37,6 +37,11 @@ contract EarlyFailTest is Test {
         // - counterexample caused by assert(x > 0): x == 0
         assert(x > 0);
     }
+
+    function check_failed_cheatcode() public {
+        // since fail() halts immediately, failed() always return true here
+        assertFalse(failed());
+    }
 }
 
 contract FoundryTest is Test {

--- a/tests/regression/test/Foundry.t.sol
+++ b/tests/regression/test/Foundry.t.sol
@@ -39,7 +39,7 @@ contract EarlyFailTest is Test {
     }
 
     function check_failed_cheatcode() public {
-        // since fail() halts immediately, failed() always return true here
+        // since fail() halts immediately, failed() always returns true here
         assertFalse(failed());
     }
 }

--- a/tests/regression/test/HalmosCheatCode.t.sol
+++ b/tests/regression/test/HalmosCheatCode.t.sol
@@ -115,6 +115,11 @@ contract HalmosCheatCodeTest is SymTest, Test {
         // NOTE: the following reverts due to the failure of the nonzero check for extcodesize(svm)
         // Dummy(address(svm)).foo();
     }
+
+    function check_enableSymbolicStorage() public {
+        // symbolic storage is not allowed for a nonexistent account
+        svm.enableSymbolicStorage(address(0xdeadbeef)); // HalmosException
+    }
 }
 
 interface Dummy {

--- a/tests/regression/test/HalmosCheatCode.t.sol
+++ b/tests/regression/test/HalmosCheatCode.t.sol
@@ -116,7 +116,25 @@ contract HalmosCheatCodeTest is SymTest, Test {
         // Dummy(address(svm)).foo();
     }
 
-    function check_enableSymbolicStorage() public {
+    function check_enableSymbolicStorage_pass(uint val) public {
+        address dummy = address(new Beep());
+        // initial value is zero
+        assertEq(vm.load(dummy, bytes32(0)), 0);
+
+        vm.store(dummy, bytes32(0), bytes32(val));
+        svm.enableSymbolicStorage(dummy);
+        // enableSymbolicStorage updates only uninitialized slots
+        assertEq(vm.load(dummy, bytes32(0)), bytes32(val));
+    }
+
+    function check_enableSymbolicStorage_fail() public {
+        address dummy = address(new Beep());
+        svm.enableSymbolicStorage(dummy);
+        // storage slots have been initialized with a symbolic value
+        assertEq(vm.load(dummy, bytes32(0)), 0); // fail
+    }
+
+    function check_enableSymbolicStorage_nonexistent() public {
         // symbolic storage is not allowed for a nonexistent account
         svm.enableSymbolicStorage(address(0xdeadbeef)); // HalmosException
     }

--- a/tests/regression/test/Store.t.sol
+++ b/tests/regression/test/Store.t.sol
@@ -47,4 +47,14 @@ contract StoreTest is Test {
         assert(c.a(key) == value);
         assert(uint(vm.load(address(c), bytes32(uint(keccak256(abi.encode(2))) + key))) == value);
     }
+
+    function check_store_nonexistent_account() public {
+        // vm.store() is not allowed for a nonexistent account
+        vm.store(address(0xdeadbeef), bytes32(0), bytes32(0)); // HalmosException
+    }
+
+    function check_load_nonexistent_account(uint slot) public {
+        // vm.load() from nonexistent accounts returns zero
+        assertEq(vm.load(address(0xdeadbeef), bytes32(slot)), 0);
+    }
 }

